### PR TITLE
EVG-5432 Support expansions in numhosts field

### PR DESCRIPTION
**Pull request by bsamek**

### CODE REVIEW

**Comment by bsamek** [comment]
This is essentially the same as https://github.com/evergreen-ci/evergreen/commit/8002c7b28664630bdcf7b040bada34c239a8a6c4, except rebased, taking into account John's work to include expansions in task spawn repro, and with `omitempty` added to the struct, which I believe is what was causing the error that caused the revert.

File: apimodels/agent_models.go [hunk 1]
```diff
@@ -75,29 +75,30 @@ type EndTaskResponse struct {
...
-	UserdataCommand string      `json:"userdata_command" plugin:"expand"`
+	AMI             string      `mapstructure:"ami" json:"ami,omitempty" plugin:"expand"`
+	Distro          string      `mapstructure:"distro" json:"distro,omitempty" plugin:"expand"`
+	EBSDevices      []EbsDevice `mapstructure:"ebs_block_device" json:"ebs_block_device,omitempty" plugin:"expand"`
+	InstanceType    string      `mapstructure:"instance_type" json:"instance_type,omitempty" plugin:"expand"`
+	Region          string      `mapstructure:"region" json:"region,omitempty" plugin:"expand"`
+	SecurityGroups  []string    `mapstructure:"security_group_ids" json:"security_group_ids,omitempty" plugin:"expand"`
+	Spot            bool        `mapstructure:"spot" json:"spot,omitempty"`
+	Subnet          string      `mapstructure:"subnet_id" json:"subnet_id,omitempty" plugin:"expand"`
+	UserdataFile    string      `mapstructure:"userdata_file" json:"-,omitempty" plugin:"expand"`
```
Comment (tychoish) [inline apimodels/agent_models.go]: I think - and omit empty are orthogonal.

File: apimodels/agent_models.go [hunk 2]
```diff
@@ -75,29 +75,30 @@ type EndTaskResponse struct {
...
+	UserdataCommand string      `json:"userdata_command" plugin:"expand,omitempty"`
```
Comment (tychoish) [inline apimodels/agent_models.go]: omit empty should g o on the json tag not the plugin tag.

File: apimodels/agent_models.go [hunk 3]
```diff
@@ -75,29 +75,30 @@ type EndTaskResponse struct {
...
-	Subnet          string      `mapstructure:"subnet_id" json:"subnet_id" plugin:"expand"`
-	UserdataFile    string      `mapstructure:"userdata_file" json:"-" plugin:"expand"`
-	UserdataCommand string      `json:"userdata_command" plugin:"expand"`
+	AMI             string      `mapstructure:"ami" json:"ami,omitempty" plugin:"expand"`
+	Distro          string      `mapstructure:"distro" json:"distro,omitempty" plugin:"expand"`
+	EBSDevices      []EbsDevice `mapstructure:"ebs_block_device" json:"ebs_block_device,omitempty" plugin:"expand"`
+	InstanceType    string      `mapstructure:"instance_type" json:"instance_type,omitempty" plugin:"expand"`
+	Region          string      `mapstructure:"region" json:"region,omitempty" plugin:"expand"`
+	SecurityGroups  []string    `mapstructure:"security_group_ids" json:"security_group_ids,omitempty" plugin:"expand"`
+	Spot            bool        `mapstructure:"spot" json:"spot,omitempty"`
```
Comment (tychoish) [inline apimodels/agent_models.go]: hrm?

File: apimodels/agent_models.go [hunk 4]
```diff
@@ -75,29 +75,30 @@ type EndTaskResponse struct {
...
-	InstanceType    string      `mapstructure:"instance_type" json:"instance_type" plugin:"expand"`
-	Region          string      `mapstructure:"region" json:"region" plugin:"expand"`
-	SecurityGroups  []string    `mapstructure:"security_group_ids" json:"security_group_ids" plugin:"expand"`
-	Spot            bool        `mapstructure:"spot" json:"spot"`
-	Subnet          string      `mapstructure:"subnet_id" json:"subnet_id" plugin:"expand"`
-	UserdataFile    string      `mapstructure:"userdata_file" json:"-" plugin:"expand"`
-	UserdataCommand string      `json:"userdata_command" plugin:"expand"`
+	AMI             string      `mapstructure:"ami" json:"ami,omitempty" plugin:"expand"`
+	Distro          string      `mapstructure:"distro" json:"distro,omitempty" plugin:"expand"`
+	EBSDevices      []EbsDevice `mapstructure:"ebs_block_device" json:"ebs_block_device,omitempty" plugin:"expand"`
```
Comment (tychoish) [inline apimodels/agent_models.go]: maybe not?

**Pull request merged**